packages/component/src/providers/CustomElements/customElements/CodeBlock.ts

```diff
@@ -0,0 +1,182 @@
+/* eslint-disable class-methods-use-this */
+import { hooks } from 'botframework-webchat-api';
+import { ReactNode, useMemo } from 'react';
+import { useRefFrom } from 'use-ref-from';
+
+import { useStyleSet } from '../../../hooks';
+import { defaultHighlightCode, HighlightCodeFn } from '../../../hooks/internal/codeHighlighter';
+import { parseDocumentFragmentFromString, useCodeHighlighter } from '../../../internal';
+import { createElementRegistryWithUpdater } from '../private/useUpdater';
+
+const { useStyleOptions, useLocalizer } = hooks;
+
+class CodeBlock extends HTMLElement {
+  static observedAttributes = ['theme', 'language'];
+
+  connected = false;
+  copyButtonElement: HTMLElement;
+  #originalFragment: DocumentFragment = undefined;
+  #updateTask?: Promise<void>;
+
+  get code() {
+    return this.querySelector('code')?.textContent ?? '';
+  }
+
+  get theme() {
+    return this.getAttribute('theme');
+  }
+
+  get language() {
+    return this.getAttribute('language');
+  }
+
+  get options() {
+    const { theme } = this;
+    return theme ? { theme } : undefined;
+  }
+
+  connectedCallback() {
+    this.connected = true;
+    this.scheduleUpdate();
+  }
+
+  disconnectedCallback() {
+    this.connected = false;
+  }
+
+  attributeChangedCallback(_name: string, oldValue: string, newValue: string) {
+    const updated = !Object.is(oldValue, newValue);
+    this.connected && updated && this.scheduleUpdate();
+  }
+
+  scheduleUpdate() {
```

Comment:
Should we make this private?

```suggestion
  #scheduleUpdate() {
```

Reply:
We use it from the outside, it is handy :smile: